gendesk.1
.\"             -*-Nroff-*-
.\"
.TH "gendesk" 1 "20 May 2024" "" ""
.SH NAME
gendesk \- generate .desktop files based on command-line arguments (or a PKGBUILD file)
.SH SYNOPSIS
.B gendesk
[flags or PKGBUILD file]
.SH DESCRIPTION
Supported PKGBUILD variables that will be included in the generated file:
.sp
.B _name
.sp
.B _genericname
.sp
.B _comment
.sp
.B _mimetype
.sp
.B _exec
.sp
.B _categories
.sp
Gendesk will try to find the correct icon from the Open Icon Library or else fall back on the default icon.
.sp
The correct application category will be guessed if not provided.
.sp.
Supported environment variables:
.sp
.B pkgname
.sp
.B pkgdesc
.sp
.SH "EXAMPLES"
.B gendesk
  Exits with error code 1 if ../PKGBUILD is not found and pkgname is not given and $pkgname is not defined. Will try to use both $pkgname and $pkgdesc.
.sp
.B gendesk PKGBUILD
  Looks for a PKGBUILD in the current directory and generates a .desktop file.
.sp
.B gendesk --pkgname test --pkgdesc test
  Generates a .desktop file for the name "test" with the description "test".
.sp
.B gendesk /home/user/archpackages/mypackage/PKGBUILD
  Generates a .desktop file from the given PKGBUILD.
.sp
A package name must be given, either by specifying a PKGBUILD file, using
\-\-pkgname or by defining a $pkgname environment variable.
.sp
Note that a "-bin", "-git", "-hg" or "-svn" suffix in a package name will be ignored, if present.
.PP
.SH OPTIONS
.TP
.B \-\-version
displays the current version information
.TP
.B \-h or \-\-help
displays brief usage information
.TP
.B \-n
don't attempt to download an icon, if missing
.TP
.B \-\-nocolor
don't use colored text
.TP
.B \-q
quiet; don't output text
.TP
.B \-f
force; overwrite files if needed
.TP
.B \-wm
generate a small desktop file for launching a window manager instead
.TP
.B \-\-pkgname
use this package name for the application (ie. emacs)
.TP
.B \-\-pkgdesc
use this package description for describing the application and try to find a suitable category
.TP
.B \-\-name
use this name for the desktop shortcut (ie. Emacs)
.TP
.B \-\-genericname
use this generic name for the desktop shortcut (ie. Text Editor)
.TP
.B \-\-comment
use this comment for the desktop shortcut (ie. Use this to edit files)
.TP
.B \-\-exec
specify an alternative executable, (ie. /usr/bin/emacs)
.TP
.B \-\-path
specify a starting path for the program, (ie. /usr/share/kotlin)
.TP
.B \-\-icon
specify an icon filename (defaults to the pkgname)
.TP
.B \-\-categories
specify categories (ie. Utility;TextEditor;)
.TP
.B \-\-terminal
specify if the application should be run in a terminal (default is false)
.TP
.B \-\-mimetypes
specify mime-types
.TP
.B \-\-startupnotify
specify if a desktop notification should occur when the application starts (default is false)
.TP
.B \-\-custom
specify an extra line (or several lines) to append at the end
.PP
.SH "WHY"
.sp
Aims to make it easy for package maintainers to add menu entries
for the applications they try to package, without including anything that
hasn't been supplied by upstream in the package itself.
.SH BUGS
Only unknown bugs so far. Bugs can be reported at https://github.com/xyproto/gendesk/issues.
.SH VERSION
1.0.8
.SH AUTHOR
.B gendesk
was written by Alexander F. Rødseth <xyproto@archlinux.org>


CHANGELOG.md
Changes from 1.0.1 to 1.0.3
---------------------------
* Flags overrides the values from a given PKGBUILD file.
* Switch a dependency from `akrennmair/gconf` to `unknwon/goconf`.
* Switch a dependency from `xyproto/term` to `xyproto/textoutput`.

Changes from 1.0.1 to 1.0.2
---------------------------
* Specify version `1.0` in generated `.desktop` files, instead of version `1.2`.

Changes from 1.0.0 to 1.0.1
---------------------------
* Fix a typo in the `--help` output.
* Update the release script.

Changes from 0.7.0 to 1.0.0
---------------------------
* Add `--icon` flag, ref #7.
* Update to the desktop-entry-spec 1.2 format (remove `Encoding` and specify `Version`), ref #8.
* Several minor changes, as suggested by the `golint` utility.
* Tested with Go 1.11.

Changes from 0.6.5 to 0.7.0
---------------------------
* Updated vendored dependencies.
* Added support for [goreleaser](https://github.com/goreleaser/goreleaser).
* Improved handling of icons, if an icon is missing.
* Minor changes and refactoring.

Changes from 0.6.4 to 0.6.5
---------------------------
* Ignore the `-svn` suffix in package names (same as for `-git`, thanks @mstraube).
* Use `text/template` for generating the `.desktop` file contents.
* Minor changes to the command line output/documentation.
* Some refactoring.
* Tested with Go 1.9.

Changes from 0.6.3 to 0.6.4
---------------------------
* Fix bug where some flags could not be overridden.

Changes from 0.6.2 to 0.6.3
---------------------------
* Will now ignore the `-git` suffix if it is part of a package name.

Changes from 0.6.1 to 0.6.2
---------------------------
* Added the possibility of having a configuration file for specifying a different URL for searching for missing icons.
* Remove the `--iconurl` flag.
* Refactored out some code to an external package.

Changes from 0.6 to 0.6.1
-------------------------
* Support for `StartupNotify=true`/`false`
* Both `--mimetype` and `--mimetypes` are allowed
* Guesses more categories than before

Changes from 0.5.5 to 0.6
-------------------------
* Added an option for generating .desktop files for launching window managers

Changes from 0.5.4 to 0.5.5
---------------------------
* Bug fix when generating .desktop files from PKGBUILD files.

Changes from 0.5.3 to 0.5.4
---------------------------
* Added a `-f` flag for overwriting files (will not overwrite without it).
* Some refactoring

Changes from 0.5.2 to 0.5.3
---------------------------
* Added a `--terminal` flag for specifying if the application should be run in a terminal.
* Some refactoring.

Changes from 0.5.1 to 0.5.2
---------------------------
* Support for additional environment variables.

Changes from 0.5.0 to 0.5.1
---------------------------
* Support for `$pkgname` and `$pkgdesc`.
* Updated the man page.
* Will try to download icons specified with `--iconurl`.

Changes from 0.4.4 to 0.5.0
---------------------------
* Command line options, no need to specify a PKGBUILD.

Changes from 0.4.3 to 0.4.4
---------------------------
* Changed the URL for searching for icons from Fedora to Open Icon Library

Changes from 0.4.2 to 0.4.3
---------------------------
* Fixed minor bug where puzzle games were not placed in the right category
* Added \_categories=()

Changes from 0.4.1 to 0.4.2
---------------------------
* Added category "Graphics;3DGraphics;" for 3D modellers
* Added category "System;" for sensor monitors
* Added category "Game;BoardGame;" for kw "board", "chess", "goban" or "chessboard"
* Added category "Office" for kw "e-book" and "ebook"
* Doesn't use ".png" by default when specifying an icon

Changes from 0.4 to 0.4.1
-------------------------
* Fixed a bug where \_name=() and \_comment=() didn't work as they should

Changes from 0.3 to 0.4
-----------------------
* Added \_genericname=()
* Added \_comment=()
* Added \_mimetype=()
* Added Type=Application
* Added category "Game;LogicGame" for keyword "puzzle"
* Added category "Game;ArcadeGame" for keyword "fighting"
* Fixed weird formatting in --help output
* Added \_custom=() for adding custom fields at the end of the .desktop file
* Glob for existing .svg icons too
* Shorter lines
* Moved functions and settings related to terminal output to a separate file

Changes from 0.2 to 0.3
-----------------------
* New flag: -q for quiet
* New flag: --nocolor for no color
* New flag: -n for not downloading anything (only generate a .desktop file)
* New flag: -q for quiet (no stdout output)
* Added \_name=('Name') to be able to specify a name that isn't only lowercase (like "ZynAddSubFX" or "jEdit")
* kw "synthesizer" is now category AudioVideo
* kw "editor" is now category TextEditor and/or Development;TextEditor
* kw "emulator" is now category "Game"
* kw "game" is now category "Game"
* kw "combat" is now be category "Game;ArcadeGame"
* kw "GPS" or "inspecting" is now category "Application;Science"
* kw "player" is now category "Application;Game;"
* kw "shooter" is now "Application;Game;ActionGame;"
* kw "roguelike" is now "Application;Game;AdventureGame;"
* kw "git" is now category Development;RevisionControl


gendesk.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	. "github.com/xyproto/textgui"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

const (
	version_string = "Desktop File Generator v.0.6.2"
)

var (
	// Global flags
	use_color = true
	verbose   = true
)

// Generate the contents for the .desktop file (for executing a window manager)
func createWindowManagerDesktopContents(name string, exec string) *bytes.Buffer {
	var buf []byte
	b := bytes.NewBuffer(buf)
	b.WriteString("[Desktop Entry]\n")
	b.WriteString("Type=XSession\n")
	b.WriteString("Exec=" + exec + "\n")
	b.WriteString("TryExec=" + exec + "\n")
	b.WriteString("Name=" + name + "\n")
	return b
}

// Generate the contents for the .desktop file (for executing a desktop application)
func createDesktopContents(name string, genericName string, comment string,
	exec string, icon string, useTerminal bool,
	categories []string, mimeTypes []string,
	startupNotify bool) *bytes.Buffer {
	var buf []byte
	b := bytes.NewBuffer(buf)
	b.WriteString("[Desktop Entry]\n")
	b.WriteString("Encoding=UTF-8\n")
	b.WriteString("Type=Application\n")
	b.WriteString("Name=" + name + "\n")
	if genericName != "" {
		b.WriteString("GenericName=" + genericName + "\n")
	}
	b.WriteString("Comment=" + comment + "\n")
	b.WriteString("Exec=" + exec + "\n")
	b.WriteString("Icon=" + icon + "\n")
	if useTerminal {
		b.WriteString("Terminal=true\n")
	} else {
		b.WriteString("Terminal=false\n")
	}
	if startupNotify {
		b.WriteString("StartupNotify=true\n")
	} else {
		b.WriteString("StartupNotify=false\n")
	}
	b.WriteString("Categories=" + strings.Join(categories, ";") + ";\n")
	if len(mimeTypes) > 0 {
		b.WriteString("MimeType=" + strings.Join(mimeTypes, ";") + ";\n")
	}
	return b
}

// Write the .desktop file as generated by createWindowManagerDesktopContents
func writeWindowManagerDesktopFile(pkgname, name, exec, custom string, force bool, o *TextOutput) {
	buf := createWindowManagerDesktopContents(name, exec)
	if custom != "" {
		// Write the custom string to the end of the .desktop file (may contain \n)
		buf.WriteString(custom + "\n")
	}
	// Check if the file exists (and that force is not enabled)
	if _, err := os.Stat(pkgname + ".desktop"); err == nil && (!force) {
		o.Err("no")
		o.Println(pkgname + ".desktop already exists. Use -f as the first argument in order to overwrite it.")
		os.Exit(1)
	}

	ioutil.WriteFile(pkgname+".desktop", buf.Bytes(), 0666)
}

// Write the .desktop file as generated by createDesktopContents
func writeDesktopFile(pkgname string, name string, comment string, exec string,
	useTerminal bool, categories string, genericName string, mimeTypes string, startupNotify bool, custom string, force bool, o *TextOutput) {
	var categoryList []string
	var mimeTypeList []string

	if len(categories) == 0 {
		categoryList = []string{"Application"}
	} else {
		categoryList = strings.Split(categories, ";")
	}
	// mimeTypeList is an empty []string, or...
	if len(mimeTypes) != 0 {
		mimeTypeList = strings.Split(mimeTypes, ";")
	}

	// mimeTypes may be empty. Disabled terminal
	// and startupnotify for now.
	buf := createDesktopContents(name, genericName, comment, exec, pkgname,
		useTerminal, categoryList, mimeTypeList, startupNotify)
	if custom != "" {
		// Write the custom string to the end of the .desktop file (may contain \n)
		buf.WriteString(custom + "\n")
	}

	// Check if the file exists (and that force is not enabled)
	if _, err := os.Stat(pkgname + ".desktop"); err == nil && (!force) {
		o.Err("no")
		o.Println(pkgname + ".desktop already exists. Use -f as the first argument to gendesk to overwrite.")
		os.Exit(1)
	}

	ioutil.WriteFile(pkgname+".desktop", buf.Bytes(), 0666)
}

// Check if a keyword appears in a package description
func keywordsInDescription(pkgdesc string, keywords []string) bool {
	for _, keyword := range keywords {
		if has(pkgdesc, keyword) {
			return true
		}
	}
	return false
}

func WriteDefaultIconFile(pkgname string, o *TextOutput) error {
	defaultIconFilename := "/usr/share/pixmaps/default.png"
	b, err := ioutil.ReadFile(defaultIconFilename)
	if err != nil {
		o.Err("could not read " + defaultIconFilename + "!")
	}
	filename := pkgname + ".png"
	err = ioutil.WriteFile(filename, b, 0666)
	if err != nil {
		o.Err("could not write icon to " + filename + "!")
	}
	return nil
}

func parseFlagsRetrieveInformation() {
}

func main() {

	var filename string
	version_help := "Show application name and version"
	nodownload_help := "Don't download anything"
	nocolor_help := "Don't use colors"
	quiet_help := "Don't output anything on stdout"
	force_help := "Overwrite .desktop files with the same name"
	windowmanager_help := "Generate a .desktop file for launching a window manager"
	pkgname_help := "The name of the package"
	pkgdesc_help := "Description of the package"
	name_help := "Name of the shortcut"
	genericname_help := "Type of application"
	comment_help := "Shortcut comment"
	exec_help := "Path to executable"
	//iconurl_help := "URL to icon"
	terminal_help := "Run the application in a terminal (default is false)"
	categories_help := "Categories, see other .desktop files for examples"
	mimetypes_help := "Mime types, see other .desktop files for examples"
	startupnotify_help := "Notifcation when the application starts (default is false)"
	custom_help := "Custom line to append at the end of the .desktop file"

	flag.Usage = func() {
		fmt.Println()
		fmt.Println(version_string)
		fmt.Println("generates .desktop files")
		fmt.Println()
		fmt.Println("Syntax: gendesk [flags] [PKGBUILD filename]")
		fmt.Println()
		fmt.Println("Possible flags:")
		fmt.Println("    --version                    " + version_help)
		fmt.Println("    -n                           " + nodownload_help)
		fmt.Println("    --nocolor                    " + nocolor_help)
		fmt.Println("    -q                           " + quiet_help)
		fmt.Println("    -f                           " + force_help)
		fmt.Println("    -wm                          " + windowmanager_help)
		fmt.Println("    --pkgname=PKGNAME            " + pkgname_help)
		fmt.Println("    --pkgdesc=PKGDESC            " + pkgdesc_help)
		fmt.Println("    --name=NAME                  " + name_help)
		fmt.Println("    --genericname=GENERICNAME    " + genericname_help)
		fmt.Println("    --comment=COMMENT            " + comment_help)
		fmt.Println("    --exec=EXEC                  " + exec_help)
		//fmt.Println("    --iconurl=ICON               " + iconurl_help)
		fmt.Println("    --terminal=[true|false]      " + terminal_help)
		fmt.Println("    --categories=CATEGORIES      " + categories_help)
		fmt.Println("    --mimetypes=MIMETYPES        " + mimetypes_help)
		fmt.Println("    --startupnotify=[true|false] " + startupnotify_help)
		fmt.Println("    --custom=CUSTOM              " + custom_help)
		fmt.Println("    --help                       This text")
		fmt.Println()
		fmt.Println("Note:")
		fmt.Println("    * Either provide a set of arguments, or a PKGBUILD file.")
		fmt.Println("    * If pkgname and pkgdesc are provided, a PKGBUILD file is not needed.")
		fmt.Println("    * \"../PKGBUILD\" is the default filename.")
		fmt.Println("    * _exec in the PKGBUILD can be used to specifiy a different executable for the")
		fmt.Println("      .desktop file. Example: _exec=('appname-gui')")
		fmt.Println("    * Split PKGBUILD packages are supported.")
		fmt.Println("    * If a .png or .svg icon is not found as a file or in the PKGBUILD, an icon")
		fmt.Println("      will be downloaded from either the download location specified in the")
		shortname := strings.Split(default_icon_search_url, "/")
		firstpart := strings.Join(shortname[:3], "/")
		fmt.Println("      configuration or from: " + firstpart)
		fmt.Println("      (This may or may not result in the icon you wished for).")
		fmt.Println("    * Categories are guessed based on keywords in the")
		fmt.Println("      package description, unless provided.")
		fmt.Println("    * Icons are assumed to be found in \"/usr/share/pixmaps/\" once installed.")
		fmt.Println()
	}

	version := flag.Bool("version", false, version_help)
	nodownload := flag.Bool("n", false, nodownload_help)
	nocolor := flag.Bool("nocolor", false, nocolor_help)
	quiet := flag.Bool("q", false, quiet_help)
	force := flag.Bool("f", false, force_help)
	windowmanager := flag.Bool("wm", false, windowmanager_help)
	givenPkgname := flag.String("pkgname", "", pkgname_help)
	givenPkgdesc := flag.String("pkgdesc", "", pkgdesc_help)
	name := flag.String("name", "", name_help)
	genericname := flag.String("genericname", "", genericname_help)
	comment := flag.String("comment", "", comment_help)
	exec := flag.String("exec", "", exec_help)
	terminal := flag.Bool("terminal", false, terminal_help)
	categories := flag.String("categories", "", categories_help)
	mimetypes := flag.String("mimetypes", "", mimetypes_help)
	mimetype := flag.String("mimetype", "", mimetypes_help)
	custom := flag.String("custom", "", custom_help)
	startupnotify := flag.Bool("startupnotify", false, startupnotify_help)
	flag.Parse()
	args := flag.Args()

	// New output. Color? Enabled?
	o := NewTextOutput(!*nocolor, !*quiet)

	if *version {
		o.Println(version_string)
		os.Exit(0)
	}

	pkgname := *givenPkgname
	pkgdesc := *givenPkgdesc
	manualIconurl := ""

	if pkgname == "" {
		if len(args) == 0 {
			if os.Getenv("pkgname") == "" {
				filename = "../PKGBUILD"
			} else {
				pkgname = os.Getenv("pkgname")
			}
		} else if len(args) >= 1 {
			// args are non-flag arguments
			filename = args[0]
		}
	}

	// Environment variables
	dataFromEnvironment(&pkgdesc, exec, name, genericname, mimetypes, comment, categories, custom)

	var pkgnames []string
	var iconurl string

	// Several fields are stored per pkgname, hence the maps
	pkgdescMap := make(map[string]string)
	execMap := make(map[string]string)
	nameMap := make(map[string]string)
	genericNameMap := make(map[string]string)
	mimeTypesMap := make(map[string]string)
	commentMap := make(map[string]string)
	categoriesMap := make(map[string]string)
	customMap := make(map[string]string)

	if filename == "" {
		// Fill in the dictionaries using the arguments
		pkgnames = []string{pkgname}
		if pkgdesc != "" {
			pkgdescMap[pkgname] = pkgdesc
		}
		if *exec != "" {
			execMap[pkgname] = *exec
		}
		if *name != "" {
			nameMap[pkgname] = *name
		}
		if *genericname != "" {
			genericNameMap[pkgname] = *genericname
		}
		if *mimetype != "" {
			mimeTypesMap[pkgname] = *mimetype
		}
		if *mimetypes != "" {
			mimeTypesMap[pkgname] = *mimetypes
		}
		if *comment != "" {
			commentMap[pkgname] = *comment
		}
		if *categories != "" {
			categoriesMap[pkgname] = *categories
		}
		if *custom != "" {
			customMap[pkgname] = *custom
		}
	} else {
		// TODO: Use a struct per pkgname instead
		parsePKGBUILD(o, filename, &iconurl, &pkgname, &pkgnames, &pkgdescMap, &execMap, &nameMap, &genericNameMap, &mimeTypesMap, &commentMap, &categoriesMap, &customMap)
	}

	// Write .desktop and .png icon for each package
	for _, pkgname := range pkgnames {
		if strings.Contains(pkgname, "-nox") ||
			strings.Contains(pkgname, "-cli") {
			// Don't bother if it's a -nox or -cli package
			continue
		}
		// TODO: Find a better way for all the if checks below
		pkgdesc, found := pkgdescMap[pkgname]
		if !found {
			// Fall back on the package name
			pkgdesc = pkgname
		}
		exec, found := execMap[pkgname]
		if !found {
			// Fall back on the package name
			exec = pkgname
		}
		name, found := nameMap[pkgname]
		if !found {
			// Fall back on the capitalized package name
			name = capitalize(pkgname)
		}
		genericName, found := genericNameMap[pkgname]
		if !found {
			// Fall back on no generic name
			genericName = ""
		}
		comment, found := commentMap[pkgname]
		if !found {
			// Fall back on pkgdesc
			comment = pkgdesc
		}
		mimeTypes, found := mimeTypesMap[pkgname]
		if !found {
			// Fall back on no mime type
			mimeTypes = ""
		}
		custom, found := customMap[pkgname]
		if !found {
			// Fall back on no custom additional lines
			custom = ""
		}
		categories, found := categoriesMap[pkgname]
		if !found {
			categories = GuessCategory(pkgdesc)
		}

		// TODO: Refactor into a function
		const nSpaces = 32
		spaces := strings.Repeat(" ", nSpaces)[:nSpaces-min(nSpaces, len(pkgname))]
		if o.IsEnabled() {
			fmt.Printf("%s%s%s%s%s ",
				o.DarkGray("["), o.LightBlue(pkgname),
				o.DarkGray("]"), spaces,
				o.DarkGray("Generating desktop file..."))
		}

		if *windowmanager {
			writeWindowManagerDesktopFile(pkgname, name, exec, custom, *force, o)
		} else {
			writeDesktopFile(pkgname, name, comment, exec, *terminal, categories, genericName, mimeTypes, *startupnotify, custom, *force, o)
		}

		if o.IsEnabled() {
			fmt.Printf("%s\n", o.DarkGreen("ok"))
		}

		// TODO: Put in a function
		// Download an icon if it's not downloaded by
		// the PKGBUILD and not there already (.png or .svg)
		pngFilenames, _ := filepath.Glob("*.png")
		svgFilenames, _ := filepath.Glob("*.svg")
		if (0 == (len(pngFilenames) + len(svgFilenames))) && !*nodownload {
			if len(pkgname) < 1 {
				o.Err("No pkgname, can't download icon")
			}
			fmt.Printf("%s%s%s%s%s ",
				o.DarkGray("["), o.LightBlue(pkgname),
				o.DarkGray("]"), spaces,
				o.DarkGray("Downloading icon..."))
			var err error
			if manualIconurl == "" {
				err = WriteIconFile(pkgname, o, *force)
			} else {
				// Default filename
				iconFilename := pkgname + ".png"
				// Get the last part of the URL, after the "/" to use as the filename
				if strings.Contains(manualIconurl, "/") {
					pos := strings.LastIndex(manualIconurl, "/")
					iconFilename = manualIconurl[pos+1:]
				}
				err = DownloadFile(manualIconurl, iconFilename, o, *force)
			}
			if err == nil {
				if o.IsEnabled() {
					fmt.Printf("%s\n", o.LightCyan("ok"))
				}
			} else {
				if o.IsEnabled() {
					fmt.Printf("%s\n", o.DarkYellow("no"))
					fmt.Printf("%s%s%s%s%s ",
						o.DarkGray("["),
						o.LightBlue(pkgname),
						o.DarkGray("]"),
						spaces,
						o.DarkGray("Using default icon instead..."))
				}
				if err := WriteDefaultIconFile(pkgname, o); (err == nil) && o.IsEnabled() {
					fmt.Printf("%s\n", o.LightPurple("yes"))
				}
			}
		}
	}
}
